### PR TITLE
Feed syncing

### PR DESCRIPTION
**Pull request by adomasven**
Adds feed syncing and makes SyncedSettings JSON for feeds more compact as discussed in #1037

### CODE REVIEW

File: test/tests/feedTest.js [hunk 1]
```diff
@@ -200,7 +200,21 @@ describe("Zotero.Feed", function() {
...
+				name: settings[0], 
+				cleanupAfter: settings[1], 
+				refreshInterval: settings[2]
+			});
+			
+			let syncedFeeds = Zotero.SyncedSettings.get(Zotero.Libraries.userLibraryID, 'feeds');
+			assert.deepEqual(syncedFeeds[url], settings);
+		});	
+		
+		it("should not store settings for feed with the old format", function* () {
```
Comment (dstillman) [inline test/tests/feedTest.js]: No need for this test, and can just call the other one "should store settings for the feed".

File: chrome/content/zotero/xpcom/data/feeds.js [hunk 2]
```diff
@@ -30,6 +30,18 @@ Zotero.Feeds = new function() {
...
+			(oldValue, newValue, conflict) => { 
+				Zotero.Feeds.restoreFromJSON(newValue, conflict);
+			}
+		);
+		
+		Zotero.Notifier.registerObserver({notify: function(event) {
+			if (event == 'finish') {
+				Zotero.Feeds.updateFeeds();
+			}
+		}}, ['sync']);
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/data/feeds.js]: Should 'syncedSettings' as the id for the observer.

File: chrome/content/zotero/xpcom/data/feeds.js [hunk 3]
```diff
@@ -259,4 +271,16 @@ Zotero.Feeds = new function() {
 		Zotero.debug("All feed updates done");
 		this.scheduleNextFeedCheck();
 	});
+	
+	// Conversion from expansive to compact format sync json
+	// TODO: Remove after beta
+	this._compactifyFeedJSON = function(json) {
+		for (let url in json) {
+			if( Object.prototype.toString.call(json[url]) === '[object Array]' ) {
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/data/feeds.js]: `Array.isArray()`?

File: chrome/content/zotero/xpcom/syncedSettings.js [hunk 4]
```diff
@@ -175,11 +209,13 @@ Zotero.SyncedSettings = (function () {
...
-				var sql = "UPDATE syncedSettings SET value=?, version=?, synced=? "
-					+ "WHERE setting=? AND libraryID=?";
-				yield Zotero.DB.queryAsync(
-					sql, [JSON.stringify(value), version, synced, setting, libraryID]
-				);
+				var sql = "UPDATE syncedSettings SET " + (version > 0 ? "version=?, " : "") + 
+					"value=?, synced=? WHERE setting=? AND libraryID=?";
+				var args = [JSON.stringify(value), synced, setting, libraryID];
+				if (version > 0) {
+					args.splice(0, 0, version)
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/syncedSettings.js]: `args.unshift(version)` (or, what we usually do, reverse order above and use `args.push(version)` here

Comment (adomasven) [inline chrome/content/zotero/xpcom/syncedSettings.js]: Can't reverse order as the last statements are `WHERE` conditions.

File: chrome/content/zotero/xpcom/data/feeds.js (hunk 2 above)
Comment (adomasven) [inline chrome/content/zotero/xpcom/data/feeds.js]: Huh?

File: chrome/content/zotero/xpcom/syncedSettings.js (hunk 4 above)
Comment (dstillman) [inline chrome/content/zotero/xpcom/syncedSettings.js]: Yeah, I usually continue the SQL and params after adding any conditional ones. But `unshift` would be fine too.

File: chrome/content/zotero/xpcom/data/feeds.js (hunk 2 above)
Comment (dstillman) [inline chrome/content/zotero/xpcom/data/feeds.js]: Additional parameter: https://github.com/zotero/zotero/blob/master/chrome/content/zotero/xpcom/notifier.js#L39

File: chrome/content/zotero/xpcom/notifier.js [hunk 5]
```diff
@@ -34,8 +34,15 @@ Zotero.Notifier = new function(){
...
 	var _queue = {};
-	
-	
+
+
+	/**
+	 * @param ref - an object with signature {notify: function(event, type, ids, extraData) {}}
+	 * @param types - an array with a list of types of events observer should be triggered on
+	 * @param id - a string to identify the observer in debug output
+	 * @param priority - lower numbers correspond to higher priority of observer execution
```
Comment (dstillman) [inline chrome/content/zotero/xpcom/notifier.js]: Missing parameter types:

`@param {Object} ref - ...`

**Comment by dstillman** [comment]
This breaks syncing of new settings:

>     ✖ [FAIL] should upload new full items and subsequent patches
      TypeError: metadata is null at
            module.set<@chrome://zotero/content/xpcom/syncedSettings.js:233:5

**Comment by adomasven** [comment]
Am on it already.

**Pull request merged**